source/text/Font.cpp

```diff
@@ -14,573 +14,823 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 		
-		int leftChars = nextChars / 2;
-		int rightChars = nextChars - leftChars;
-		int nextWidth = WidthRawString((str.substr(0, leftChars) + str.substr(str.size() - rightChars)).c_str(), '.');
-		bool nextWorks = (nextWidth <= width);
-		if(prevWorks != nextWorks && abs(nextChars - prevChars) == 1)
-		{
-			if(prevWorks)
-			{
-				leftChars = prevChars / 2;
-				rightChars = prevChars - leftChars;
-				width = prevWidth + widthEllipses;
-			}
-			else
-				width = nextWidth + widthEllipses;
-			return str.substr(0, leftChars) + "..." + str.substr(str.size() - rightChars);
-		}
+		// Input: Texture coordinate from the vertex shader.
+		"in vec2 texCoord;\n"
 		
-		prevChars = nextChars;
-		prevWidth = nextWidth;
-	}
-	width = firstWidth;
-	return str;
+		// Output: Color for the screen.
+		"out vec4 finalColor;\n"
+		
+		"void main() {\n"
+		"  finalColor = color * texture(tex, texCoord);\n"
+		"}\n";
+	
+	shader = Shader(vertexCode, fragmentCode);
+	scaleI = shader.Uniform("scale");
+	centerI = shader.Uniform("center");
+	sizeI = shader.Uniform("size");
+	colorI = shader.Uniform("color");
+	
+	// The texture always comes from texture unit 0.
+	glUseProgram(shader.Object());
+	glUniform1i(shader.Uniform("tex"), 0);
+	glUseProgram(0);
+	
+	// Create the VAO and VBO.
+	glGenVertexArrays(1, &vao);
+	glBindVertexArray(vao);
+	
+	glGenBuffers(1, &vbo);
+	glBindBuffer(GL_ARRAY_BUFFER, vbo);
+	
+	// Triangle strip.
+	GLfloat vertexData[] = {
+		-.5f, -.5f,
+		-.5f,  .5f,
+		 .5f, -.5f,
+		 .5f,  .5f
+	};
+	glBufferData(GL_ARRAY_BUFFER, sizeof(vertexData), vertexData, GL_STATIC_DRAW);
+	
+	glEnableVertexAttribArray(shader.Attrib("vert"));
+	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), nullptr);
+	
+	// Unbind the VBO and VAO.
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+	
+	// We must update the screen size next time we draw.
+	screenWidth = 1;
+	screenHeight = 1;
+	viewWidth = 1;
+	viewHeight = 1;
+}
+
+
+
+int Font::ViewWidth(const DisplayText &text) const
+{
+	if(text.GetText().empty())
+		return 0;
+	
+	const RenderedText &renderedText = Render(text);
+	if(!renderedText.texture)
+		return 0;
+	return renderedText.width;
+}
+
+
+
+double Font::ViewFromTextX(double x) const
+{
+	return x * viewWidth / screenWidth;
+}
+
+
+
+double Font::ViewFromTextY(double y) const
+{
+	return y * viewHeight / screenHeight;
+}
+
+
+
+int Font::ViewFromTextX(int x) const
+{
+	return floor(static_cast<double>(x * viewWidth + screenWidth / 2.0) / screenWidth);
+}
+
+
+
+int Font::ViewFromTextY(int y) const
+{
+	return floor(static_cast<double>(y * viewHeight + screenHeight / 2.0) / screenHeight);
+}
+
+
+
+int Font::ViewFromTextCeilX(int x) const
+{
+	return ceil(static_cast<double>(x * viewWidth) / screenWidth);
+}
+
+
+
+int Font::ViewFromTextCeilY(int y) const
+{
+	return ceil(static_cast<double>(y * viewHeight) / screenHeight);
+}
+
+
+
+int Font::ViewFromTextFloorX(int x) const
```

Comment:
There's a lot of similarly named methods here, and from their names it's really not clear whether the implementation satisfies the caller's expectation:

I see `ViewFromTextX(int)`, which calls floor on the final result of `x * viewWidth + 1/2 screenWidth`,  divided by `screenWidth` (so `x * viewWidth / screenWidth + .5`), and then I see `ViewFromTextFloorX`, which _also_ calls floor on the final result of `x * viewWidth / screenWidth`, but this time without adding the .5 before calling `floor`.

We should clean these up / rename them, so it's more readily apparent what the purposes are. I would also suggest we group them by X vs Y, so the difference in implementations is easier to review.

Reply:
ViewFromTextX(int x) rounds into the nearest value, but rounds toward plus infinity if x has the same distance to two nearest integers. It should name ViewFromTextNearestX()?

Rounding toward zero makes the objects move to center on average. We may notice it when showing the preferences panel.